### PR TITLE
Form validation

### PR DESCRIPTION
**Pull request by williamcosta**
**DESCRIPTION:**

Add validations to form and enhancements.

**CHANGELOG:**

* Add validates to required fields;
* Add validates to empty fields;
* Add custom validates to zipcode, cellphone, minLength and maxLenght;
* Add error styles to inputs;

### Please ignore `.snap` files

**PRINTSCREEN:**

#### First Step invalidated
screen shot 2018-05-08 at 16 23 35

#### Second Step invalidated
screen shot 2018-05-08 at 16 23 53

### CODE REVIEW

File: spec/components/Form/__snapshots__/index.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,300 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (williamcosta) [inline spec/components/Form/__snapshots__/index.spec.js.snap]: Esses arquivos de `Snapshot` são criados automaticamente. Não sendo necessário fazer o review dos mesmos. 😬

File: spec/components/Form/index.spec.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
+import React from 'react';
+import renderer from 'react-test-renderer';
+import Form from '../../../src/components/Form';
+import { form } from '../../../src/form.json';
+import { enzymeConfig, shallow } from '../../enzymeConfig';
+import fillFormFields from '../../helper';
+
+enzymeConfig();
+
+function createNodeMock(element) {
```
Comment (williamcosta) [inline spec/components/Form/index.spec.js]: Precisamos criar esse mock para os input que vão receber algum tipo de interação por conta do React.

File: spec/helper.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+const fillFormFields = (steps) => {
```
Comment (williamcosta) [inline spec/helper.js]: Criamos esse método para mudarmos o valor dos fields do JSON para fazermos os testes de update das steps do form.

File: src/components/Form/validation.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import errorMessage from '../../../src/helpers/errorMessages';
+
+export const isEmpty = (value) => {
+  const regex = /^\s*$/;
+  return regex.test(value.trim());
+};
+
+export const isMinLength = (text, length) => {
```
Comment (williamcosta) [inline src/components/Form/validation.js]: Criamos esse método para validarmos o `minlength` do input. Já que essa validação nativa é feita pelo HTML.

**Comment by felipeorlando** [comment]
English, please.

File: spec/components/Form/__snapshots__/index.spec.js.snap [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+exports[`Form renders defaultProps 1`] = `
```
Comment (nathpaiva) [inline spec/components/Form/__snapshots__/index.spec.js.snap]: ❤️

File: spec/components/Form/validation.spec.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
+import { isEmpty, isMinLength, isValidEmail, isValidZipcode, isValidCellPhone, validateField, validateStep } from '../../../src/components/Form/validation';
+import { form } from '../../../src/form.json';
+import fillFormFields from '../../helper';
+import errorMessages from '../../../src/helpers/errorMessages';
+
```
Comment (nathpaiva) [inline spec/components/Form/validation.spec.js]: 👏

File: src/components/Field.jsx [hunk 7]
```diff
@@ -20,14 +22,22 @@ export default class Field extends Component {
     this.spanStyle = 'form__message form__message--invalid space-element-small error';
   }
 
+  get style() {
```
Comment (nathpaiva) [inline src/components/Field.jsx]: 👍

File: src/components/Step.jsx [hunk 8]
```diff
@@ -36,18 +43,28 @@ export default class Step extends Component {
                 key={`field-${index}`}
                 label={item.title}
                 id={item.id}
-                errorMessage='Required field'>
-
-                { Factory.getComponent(item, index) }
+                errorMessage={item.errorMessage} >
```
Comment (nathpaiva) [inline src/components/Step.jsx]: could you remove the space between `errorMessage={item.errorMessage}` and `>`? 😁

File: spec/jsdomConfig.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import jsdom from 'jsdom';
```
Comment (williamcosta) [inline spec/jsdomConfig.js]: Added `jsdom` to resolve problem with *DOM Environment* in `mount` method.

http://airbnb.io/enzyme/docs/guides/jsdom.html

File: spec/App.spec.js [hunk 10]
```diff
@@ -2,6 +2,9 @@ import React from 'react';
 import ReactDOM from 'react-dom';
 import App from '../src/App';
 import { form } from '../src/form.json';
+import jsdomConfig from './jsdomConfig';
+
+jsdomConfig();
```
Comment (iondrimba) [inline spec/App.spec.js]: Nice @williamcosta 👏

File: spec/components/Form/__snapshots__/index.spec.js.snap (hunk 1 above)
Comment (Eliada) [reply spec/components/Form/__snapshots__/index.spec.js.snap]: If they are created automatically, do they need to be committed?
Doesn't the build process generates them o production (like `package.json`)?

Comment (iondrimba) [reply spec/components/Form/__snapshots__/index.spec.js.snap]: Yes, is recommended to keep these snapshots versioned. 
`Commit snapshots and review them as part of your regular code review process. This means treating snapshots as you would any other type of test or code in your project.`

https://facebook.github.io/jest/docs/en/snapshot-testing.html#1-treat-snapshots-as-code

File: spec/components/Breadcrumb.spec.js [hunk 11]
```diff
@@ -16,36 +16,67 @@ describe('Breadcrumb', () => {
...
-    expect(result).toEqual('form__steps-item--active');
-    expect(secondResult).toEqual('');
-  });
+        expect(result).toEqual('form__steps-item--active');
+      })
 
-  describe('when second step is active', () => {
-    it('expects the second breadcrumb to be active', () => {
-      const component = shallow(<Breadcrumb active={1} steps={form.steps} />);
+      it('returns empty string', ()=> {
```
Comment (Eliada) [inline spec/components/Breadcrumb.spec.js]:
```js
it('returns no active class', ()=>) {
```
or
```js
it('does not return any class', ()=>) {
```

Comment (iondrimba) [reply spec/components/Breadcrumb.spec.js]: Good suggestion @Eliada 👍

File: spec/components/Field.spec.js [hunk 12]
```diff
@@ -8,17 +8,9 @@ enzymeConfig();
...
-
-    expect(tree).toMatchSnapshot();
-  });
-
-  it('renders custom props', () => {
     const component = renderer.create(
-      <Field label={'test'} id={'test'} errorMessage={'Erro!'}>
-        Bora pra action
+      <Field label={'test'} id={'test'} errorMessage={'Erro!'} value={'test value'}>
+        <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => { }} />
```
Comment (rafaeldev) [inline spec/components/Field.spec.js]: The curly brackets need has this space between them? 💅

Comment (iondrimba) [reply spec/components/Field.spec.js]: 🦅 👁

**Pull request merged**